File: 백준/Gold/17352. 여러분의 다리가 되어 드리겠습니다！/여러분의 다리가 되어 드리겠습니다！.py
# 입력 속도 개선
import sys
input = sys.stdin.readline

n = int(input())
parents = list(range(n+1)) # 부모 노드 저장

# 부모 노드 찾기
def find(x):
    if x == parents[x]:
        return x
    parents[x] = find(parents[x])
    return parents[x]

# 두 섬 중 작은 숫자를 기준으로 같은 부모를 갖게 함
def union(x,y):
    x = find(x)
    y = find(y)

    if x < y:
        parents[y] = x
    else:
        parents[x] = y

# 다리가 연결되어 있는 두 섬은 같은 부모를 갖게 함
for _ in range(n-2):
    island1, island2 = map(int,input().split())
    union(island1, island2)

# 1번 섬이 포함된 집합과 부모 노드가 다른 섬 찾아 정답 도출
# 작은 숫자가 부모 노드가 되도록 설정하였음으로 
# 1번이 1번이 포합된 집합의 부모 노드가 됨
for i in range(2,n+1):
    if find(1) != find(i):
        print(1,i)
        break

File: 백준/Gold/17352. 여러분의 다리가 되어 드리겠습니다！/여러분의 다리가 되어 드리겠습니다！.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const N=Number(input[iter++])
  if (N===2){
    console.log('1 2')
    process.exit()
  }
  const root=[...new Array(N+1)].map(item=>-1)
  const info={}

  function update(value,idx){
    if (value in info) info[value].push(idx)
    else info[value]=[idx]
  }

  for(let i=0;i<N-2;i++){
    const [a,b]=input[iter++].split(' ').map(item=>Number(item))
    if (root[a]===-1){
      if (root[b]===-1){
        const Min=Math.min(a,b)
        root[a]=Min
        root[b]=Min
        update(Min,a)
        update(Min,b)
      } else {
        root[a]=root[b]
        update(root[b],a)
      }
    } else {
      if (root[b]===-1){
        root[b]=root[a]
        update(root[a],b)
      } else {
        if (root[a]===root[b]) continue
        const Min=Math.min(root[a],root[b])
        const Max=Math.max(root[a],root[b])
        for(let i=0;i<info[Max].length;i++){
          const idx=info[Max][i]
          root[idx]=Min
          update(Min,idx)
        }
      }
    }
  }

  const first=root[1]
  for(let i=2;i<=N;i++){
    if (first!==root[i]){
      console.log(`${1} ${i}`)
      break
    }
  }
  process.exit()
})
